pkg/usagestats/reporter.go

```diff
@@ -0,0 +1,160 @@
+	"github.com/google/uuid"
+	"github.com/grafana/agent/pkg/config"
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/dskit/multierror"
+	"github.com/grafana/loki/pkg/util/build"
+)
+
+const (
+	// File name for the agent seed file
+	agentSeedFileName = "agent_seed.json"
+)
+
+var (
+	reportCheckInterval = time.Minute
+	reportInterval      = 1 * time.Hour
+)
+
+// Reporter holds the agent seed information and sends report of usage
+type Reporter struct {
+	logger log.Logger
+	cfg    *config.Config
+
+	agentSeed  *AgentSeed
+	lastReport time.Time
+}
+
+// NewReporter creates a Reporter that will send periodically reports to grafana.com
+func NewReporter(logger log.Logger, cfg *config.Config) (*Reporter, error) {
+	r := &Reporter{
+		logger: logger,
+		cfg:    cfg,
+	}
+	return r, nil
+}
+
+func (rep *Reporter) init(ctx context.Context) error {
+	if fileExists(agentSeedFileName) {
+		seed, err := rep.readSeedFile()
+		rep.agentSeed = seed
+		return err
+	}
+	rep.agentSeed = &AgentSeed{
+		UID:               uuid.NewString(),
+		PrometheusVersion: build.GetVersion(),
+		CreatedAt:         time.Now(),
+	}
+	return rep.writeSeedFile(*rep.agentSeed)
+}
+
+func fileExists(path string) bool {
+	_, err := os.Stat(path)
+	return !errors.Is(err, os.ErrNotExist)
+}
+
+// readSeedFile reads the agent seed file
+func (rep *Reporter) readSeedFile() (*AgentSeed, error) {
+	data, err := ioutil.ReadFile(agentSeedFileName)
+	if err != nil {
+		return nil, err
+	}
+	seed, err := JSONCodec.Decode(data)
+	if err != nil {
+		return nil, err
+	}
+	return seed.(*AgentSeed), nil
+}
+
+// writeSeedFile writes the agent seed file
+func (rep *Reporter) writeSeedFile(seed AgentSeed) error {
+	data, err := JSONCodec.Encode(seed)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(agentSeedFileName, data, 0644)
+}
+
+// Start inits the reporter seed and start sending report for every interval
+func (rep *Reporter) Start(ctx context.Context) error {
+	level.Info(rep.logger).Log("msg", "running usage stats reporter")
+	err := rep.init(ctx)
+	if err != nil {
+		level.Info(rep.logger).Log("msg", "failed to init seed", "err", err)
+		return err
+	}
+
+	// check every minute if we should report.
+	ticker := time.NewTicker(reportCheckInterval)
+	defer ticker.Stop()
+
+	// find  when to send the next report.
+	next := nextReport(reportInterval, rep.agentSeed.CreatedAt, time.Now())
+	if rep.lastReport.IsZero() {
+		// if we never reported assumed it was the last interval.
+		rep.lastReport = next.Add(-reportInterval)
+	}
+	for {
+		select {
+		case <-ticker.C:
+			now := time.Now()
+			if !next.Equal(now) && now.Sub(rep.lastReport) < reportInterval {
+				continue
+			}
+			level.Info(rep.logger).Log("msg", "reporting cluster stats", "date", time.Now())
+			if err := rep.reportUsage(ctx, next); err != nil {
+				level.Info(rep.logger).Log("msg", "failed to report usage", "err", err)
+				continue
+			}
+			rep.lastReport = next
+			next = next.Add(reportInterval)
```

Comment:
What's the purpose of this logic around `next? I have a feeling that (on Loki at least) it might be around its clustering logic, where a node might pick up the task left over from a previous node.

Since we're now only using a single reporter and no tickers will be running concurrently, I don't think that we'll ever hit that `continue` branch, right?

Reply:
That's a good question! I think the logic behind this besides clustering logic is to guarantee that the report will be send every `reportInterval`. In this case, even if you restart the agent it will guarantee that the report only will be send every hour.

Reply:
yes, the logic is trying to ensure that the reports are sent consistently regardless of the agent itself being restarted.